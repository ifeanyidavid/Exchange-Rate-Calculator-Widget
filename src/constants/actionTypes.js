export const FETCH_RATES_START = "FETCH_RATES_START";
export const FETCH_RATES_ERROR = "FETCH_RATES_ERROR";
export const FETCH_RATES_DONE = "FETCH_RATES_DONE";
export const UPDATE_SOURCE_CURRENCY = "UPDATE_SOURCE_CURRENCY";
export const UPDATE_DESTINATION_CURRENCY = "UPDATE_DESTINATION_CURRENCY";
export const EXCHANGE_CURRENCY = "EXCHANGE_CURRENCY";
