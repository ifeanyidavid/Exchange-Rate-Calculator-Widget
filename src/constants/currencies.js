export const currencies = {
  EUR: { id: 0, value: "EUR" },
  GBP: { id: 1, value: "GBP" },
  USD: { id: 2, value: "USD" }
};
