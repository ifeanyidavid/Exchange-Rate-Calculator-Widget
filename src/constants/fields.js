export const fieldNames = {
  sourceCurrency: "sourceCurrency",
  destinationCurrency: "destinationCurrency",
  sourceAmount: "sourceAmount",
  destinationAmount: "destinationAmount",
  balance: "balance",
  symbol: "symbol",
  description: "description",
  currency: "currency",
  id: "id",
  error: "error",
  loading: "loading",
  rates: "rates",
};
