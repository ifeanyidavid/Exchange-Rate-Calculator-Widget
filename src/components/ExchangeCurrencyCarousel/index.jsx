import React from "react";
import PropTypes from "prop-types";
import Slider from "react-slick";
import "slick-carousel/slick/slick.css";
import "slick-carousel/slick/slick-theme.css";
import CurrencyEntry from "../CurrencyEntry";
import { currencies } from "../../constants/currencies";

function ExchangeCurrencyCarousel(props) {
  const {
    currency,
    pockets,
    name,
    value,
    onChange,
    onSlide,
    showExchangeRate,
    destinationRate,
    recalculateOnSlide,
    ratesFetched
  } = props;
  const settings = {
    dots: true,
    infinite: true,
    speed: 500,
    slidesToShow: 1,
    slidesToScroll: 1,
    initialSlide: currencies[currency].id,
    swipeToSlide: true,
    responsive: [
      {
        breakpoint: 1024,
        settings: {
          slidesToShow: 1,
          slidesToScroll: 1,
          infinite: true,
          dots: true
        }
      },
      {
        breakpoint: 600,
        settings: {
          slidesToShow: 1,
          slidesToScroll: 1,
          initialSlide: 2
        }
      },
      {
        breakpoint: 480,
        settings: {
          slidesToShow: 1,
          slidesToScroll: 1
        }
      }
    ]
  };

  const handleSlide = index => {
    recalculateOnSlide();
    onSlide(index);
  };

  return (
    <Slider
      afterChange={handleSlide}
      {...settings}
      style={{ marginBottom: "30px" }}
    >
      {pockets.map(pocket => (
        <CurrencyEntry
          key={pocket.id}
          id={pocket.id}
          currency={pocket.currency}
          balance={pocket.balance}
          symbol={pocket.symbol}
          name={name}
          value={value}
          onChange={onChange}
          showExchangeRate={showExchangeRate}
          destinationRate={destinationRate}
          ratesFetched={ratesFetched}
        />
      ))}
    </Slider>
  );
}

export default ExchangeCurrencyCarousel;

ExchangeCurrencyCarousel.propTypes = {
  currency: PropTypes.string.isRequired,
  pockets: PropTypes.array.isRequired,
  name: PropTypes.string.isRequired,
  value: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  onSlide: PropTypes.func.isRequired,
  showExchangeRate: PropTypes.bool.isRequired,
  destinationRate: PropTypes.string,
  recalculateOnSlide: PropTypes.func.isRequired,
  ratesFetched: PropTypes.func,
};
