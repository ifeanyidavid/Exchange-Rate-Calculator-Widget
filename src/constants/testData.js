import { currencies } from "./currencies";
import { fieldNames } from "./fields";

export const testData = {
  rates: [
    {
      rates: { USD: 1.1075, GBP: 0.84195 },
      base: "EUR",
      date: "2019-12-09"
    },
    {
      rates: { EUR: 1.1877189857, USD: 1.3153987766 },
      base: "GBP",
      date: "2019-12-09"
    },
    {
      rates: { EUR: 0.9029345372, GBP: 0.7602257336 },
      base: "USD",
      date: "2019-12-09"
    }
  ],
  pocketsReducer: {
    transactionHistory: [
      {
        id: 0,
        description: `Exchanged to EUR`,
        time: new Date().toLocaleTimeString("en-GB"),
        sourceAmount: 23,
        destinationAmount: 42,
        sourceCurrency: "USD",
        destinationCurrency: "EUR"
      }
    ],
    pockets: {
      [currencies.EUR.value]: {
        [fieldNames.id]: currencies.EUR.id,
        [fieldNames.currency]: currencies.EUR.value,
        [fieldNames.balance]: 300,
        [fieldNames.symbol]: "€",
        [fieldNames.description]: "Euro"
      },
      [currencies.GBP.value]: {
        [fieldNames.id]: currencies.GBP.id,
        [fieldNames.currency]: currencies.GBP.value,
        [fieldNames.balance]: 300,
        [fieldNames.symbol]: "£",
        [fieldNames.description]: "British Pound"
      },
      [currencies.USD.value]: {
        [fieldNames.id]: currencies.USD.id,
        [fieldNames.currency]: currencies.USD.value,
        [fieldNames.balance]: 300,
        [fieldNames.symbol]: "$",
        [fieldNames.description]: "American Dollar"
      }
    }
  },
  sourceAmount: 23,
  destinationAmount: 54,
  sourceCurrency: "USD",
  destinationCurrency: "EUR",
  error: new Error("Network disconnected"),
  slideIndex: 1,
  pocketValues: [
    {
      [fieldNames.id]: currencies.EUR.id,
      [fieldNames.currency]: currencies.EUR.value,
      [fieldNames.balance]: 300,
      [fieldNames.symbol]: "€",
      [fieldNames.description]: "Euro"
    },
    {
      [fieldNames.id]: currencies.GBP.id,
      [fieldNames.currency]: currencies.GBP.value,
      [fieldNames.balance]: 300,
      [fieldNames.symbol]: "£",
      [fieldNames.description]: "British Pound"
    },
    {
      [fieldNames.id]: currencies.USD.id,
      [fieldNames.currency]: currencies.USD.value,
      [fieldNames.balance]: 300,
      [fieldNames.symbol]: "$",
      [fieldNames.description]: "American Dollar"
    }
  ]
};
