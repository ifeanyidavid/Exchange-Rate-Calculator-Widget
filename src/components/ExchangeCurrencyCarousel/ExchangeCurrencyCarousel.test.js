import React from "react";
import Adapter from "enzyme-adapter-react-16";
import { shallow, configure } from "enzyme";
import toJson from "enzyme-to-json";
import ExchangeCurrencyCarousel from ".";
import { fieldNames } from "../../constants/fields";
import { currencies } from "../../constants/currencies";

configure({ adapter: new Adapter() });

describe("ExchangeCurrencyCarousel component", () => {
  it("Renders correctly", () => {
    const pockets = [
      {
        [fieldNames.id]: currencies.EUR.id,
        [fieldNames.currency]: currencies.EUR.value,
        [fieldNames.balance]: 300,
        [fieldNames.symbol]: "€",
        [fieldNames.description]: "Euro"
      },
      {
        [fieldNames.id]: currencies.GBP.id,
        [fieldNames.currency]: currencies.GBP.value,
        [fieldNames.balance]: 300,
        [fieldNames.symbol]: "£",
        [fieldNames.description]: "British Pound"
      },
      {
        [fieldNames.id]: currencies.USD.id,
        [fieldNames.currency]: currencies.USD.value,
        [fieldNames.balance]: 300,
        [fieldNames.symbol]: "$",
        [fieldNames.description]: "American Dollar"
      }
    ];
    const props = {
      currency: "USD",
      pockets,
      name: "source",
      value: "20.40",
      onChange: jest.fn(),
      onSlide: jest.fn(),
      showExchangeRate: true,
      destinationRate: "1.456",
      recalculateOnSlide: jest.fn(),
      ratesFetched: jest.fn()
    };
    // wrapper instance around rendered output
    const wrapper = shallow(<ExchangeCurrencyCarousel {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
