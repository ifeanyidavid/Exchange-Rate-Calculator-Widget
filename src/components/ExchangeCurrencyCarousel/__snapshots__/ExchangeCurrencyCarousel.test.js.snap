// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExchangeCurrencyCarousel component Renders correctly 1`] = `
<Slider
  afterChange={[Function]}
  dots={true}
  infinite={true}
  initialSlide={2}
  responsive={
    Array [
      Object {
        "breakpoint": 1024,
        "settings": Object {
          "dots": true,
          "infinite": true,
          "slidesToScroll": 1,
          "slidesToShow": 1,
        },
      },
      Object {
        "breakpoint": 600,
        "settings": Object {
          "initialSlide": 2,
          "slidesToScroll": 1,
          "slidesToShow": 1,
        },
      },
      Object {
        "breakpoint": 480,
        "settings": Object {
          "slidesToScroll": 1,
          "slidesToShow": 1,
        },
      },
    ]
  }
  slidesToScroll={1}
  slidesToShow={1}
  speed={500}
  style={
    Object {
      "marginBottom": "30px",
    }
  }
  swipeToSlide={true}
>
  <CurrencyEntry
    balance={300}
    currency="EUR"
    destinationRate="1.456"
    id={0}
    key="0"
    name="source"
    onChange={[MockFunction]}
    ratesFetched={[MockFunction]}
    showExchangeRate={true}
    symbol="€"
    value="20.40"
  />
  <CurrencyEntry
    balance={300}
    currency="GBP"
    destinationRate="1.456"
    id={1}
    key="1"
    name="source"
    onChange={[MockFunction]}
    ratesFetched={[MockFunction]}
    showExchangeRate={true}
    symbol="£"
    value="20.40"
  />
  <CurrencyEntry
    balance={300}
    currency="USD"
    destinationRate="1.456"
    id={2}
    key="2"
    name="source"
    onChange={[MockFunction]}
    ratesFetched={[MockFunction]}
    showExchangeRate={true}
    symbol="$"
    value="20.40"
  />
</Slider>
`;
